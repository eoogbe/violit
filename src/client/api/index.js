/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export { load, loadAll } from './load';
export { default as InvalidResponseError } from './InvalidResponseError';
export { default as checkExists } from './checkExists';
export { default as checkPermissions } from './checkPermissions';
export { default as href } from './href';
export { default as sendCreate } from './sendCreate';
export { default as sendDelete } from './sendDelete';
export { default as sendPost } from './sendPost';
