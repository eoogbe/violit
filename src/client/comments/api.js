/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as R from 'ramda';
import {
  checkPermissions,
  href,
  loadAll,
  sendCreate,
  sendPost,
} from 'client/api';

/** Converts a comment from the API format to the view format. */
const convertComment = ({ links, ...comment }) => R.compose(
  R.assoc('canDelete', !!links.deleteAction),
  R.pick(['id', 'dateCreated', 'dateDeleted', 'text', 'threadSlug', 'author'])
)(comment);

/** Loads comments from the server. */
export const loadComments = async (threadSlug) => {
  const { data } = await loadAll(href(`/threads/${threadSlug}/comments`));
  return data.map(convertComment);
};

/**
 * Sends the specified `comment` to the server to be created.
 */
export const sendCreateComment = async (comment) => {
  const url = href(`/threads/${comment.threadSlug}/comments`);
  const { body } = await sendCreate(url, ['id', 'text'], comment);
  return convertComment(body);
};

/** Checks if the current user can create a comment. */
export const checkCanCreateComment = async (threadSlug) => {
  const url = href(`/threads/${threadSlug}/comments`);
  const methods = await checkPermissions(url);
  return methods.includes('POST');
};

/** Sends a comment to be deleted on the server. */
export const sendDeleteComment = async (commentId) => {
  await sendPost(href(`/comments/${commentId}/delete`));
  return null;
};
