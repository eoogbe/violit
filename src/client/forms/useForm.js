/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import camelCase from 'camel-case';
import * as R from 'ramda';
import { useCallback, useEffect, useMemo, useRef, useState } from 'react';

/** Hook that adds state for a form with controlled components. */
export default (initialValues, validate, options) => {
  const { commitOnChange } = options || {};

  const prevValues = useRef(initialValues);
  const preventDoubleSubmit = useRef(false);
  const [values, setValues] = useState(initialValues);
  const [errors, setErrors] = useState({});
  const [dirtied, setDirtied] = useState(R.map(R.F, initialValues));
  const [submitting, setSubmitting] = useState(false);

  const unmounted = useRef(false);
  useEffect(() => () => {
    unmounted.current = true;
  }, []);

  const isDirty = useMemo(() => {
    return Object.values(dirtied).includes(true);
  }, [dirtied]);

  const hasError = useMemo(() => {
    return Object.values(errors).some((error) => !!error);
  }, [errors]);

  const validateFields = useCallback(async () => {
    if (!validate) {
      return {};
    }

    const errors = await validate(values) || {};
    if (!unmounted.current) {
      setErrors(errors);
    }

    return errors;
  }, [validate, values]);

  useEffect(() => {
    if (values === prevValues.current) {
      return;
    }

    validateFields();
    prevValues.current = values;
  }, [values, validateFields]);

  const reset = useCallback(() => {
    if (unmounted.current) {
      return;
    }

    prevValues.current = initialValues;
    preventDoubleSubmit.current = false;
    setValues(initialValues);
    setErrors({});
    setDirtied(R.map(R.F, initialValues));
    setSubmitting(false);
  }, [initialValues]);

  const handleInput = useCallback(({ target }) => {
    const { name, value } = target;
    setValues(R.assoc(camelCase(name), value));
  }, []);

  const handleCommit = useCallback(({ target }) => {
    const { value } = target;
    const name = camelCase(target.name);
    setValues(R.assoc(name, value));
    setDirtied(R.assoc(name, true));
  }, []);

  const handleChange = useCallback((event) => {
    handleInput(event);
    if (commitOnChange) {
      handleCommit(event);
    }
  }, [handleInput, handleCommit, commitOnChange]);

  const handleSubmit = useCallback((onSubmit) => async (event) => {
    event.preventDefault();

    if (preventDoubleSubmit.current) {
      return;
    }

    if (!unmounted.current) {
      preventDoubleSubmit.current = true;
      setSubmitting(true);
    }

    const fieldErrors = await validateFields();
    if (Object.values(fieldErrors).every((error) => !error)) {
      const submitErrors = await onSubmit(values) || {};
      if (!unmounted.current) {
        setErrors(submitErrors);
      }
    }

    if (!unmounted.current) {
      preventDoubleSubmit.current = false;
      setSubmitting(false);
    }
  }, [validateFields, values]);

  const eventHandlers = {
    handleChange,
    handleBlur: handleCommit,
    handleSubmit,
  };

  return [
    values,
    errors,
    dirtied,
    eventHandlers,
    { submitting, isDirty, hasError, reset, validateFields },
  ];
};
