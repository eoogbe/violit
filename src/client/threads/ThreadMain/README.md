```jsx
import { MemoryRouter } from 'react-router-dom';
import * as THREADS from '__fixtures__/threadsFixture';

<MemoryRouter>
  <ThreadMain
    id="thread1"
    slug={THREADS.thread1.slug}
    author={THREADS.thread1.author}
    dateCreated={THREADS.thread1.dateCreated}
    headline={THREADS.thread1.headline}
    articleBody={THREADS.thread1.articleBody}
    createComment={() => {}}
    loggedIn
    canCreateComment
    canDelete
  />
</MemoryRouter>
```

Without article body:

```jsx
import { MemoryRouter } from 'react-router-dom';
import * as THREADS from '__fixtures__/threadsFixture';

<MemoryRouter>
  <ThreadMain
    id="thread1"
    slug={THREADS.thread1.slug}
    author={THREADS.thread1.author}
    dateCreated={THREADS.thread1.dateCreated}
    headline={THREADS.thread1.headline}
    createComment={() => {}}
    loggedIn
    canCreateComment
    canDelete
  />
</MemoryRouter>
```

When deleted:

```jsx
import { MemoryRouter } from 'react-router-dom';
import * as THREADS from '__fixtures__/threadsFixture';

<MemoryRouter>
  <ThreadMain
    id="thread1"
    slug={THREADS.thread1.slug}
    author={THREADS.thread1.author}
    dateCreated={THREADS.thread1.dateCreated}
    dateDeleted={THREADS.thread1.dateDeleted}
    headline={THREADS.thread1.headline}
    articleBody={THREADS.thread1.articleBody}
    createComment={() => {}}
  />
</MemoryRouter>
```

When cannot create comment:

```jsx
import { MemoryRouter } from 'react-router-dom';
import * as THREADS from '__fixtures__/threadsFixture';

<MemoryRouter>
  <ThreadMain
    id="thread1"
    slug={THREADS.thread1.slug}
    author={THREADS.thread1.author}
    dateCreated={THREADS.thread1.dateCreated}
    headline={THREADS.thread1.headline}
    articleBody={THREADS.thread1.articleBody}
    createComment={() => {}}
  />
</MemoryRouter>
```
