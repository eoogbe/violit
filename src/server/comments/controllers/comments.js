/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

const R = require('ramda');
const { coll, href, parseRequestBody } = require('../../routing');
const policies = require('../policies');
const addCommentItems = require('./addCommentItems');
const createCommentCollection = require('./createCommentCollection');

/** Creates the Collection+JSON representation of the comments resource. */
const createCommentsCollection = ({ req, res, comments }) => {
  const { credentials, routes } = res.locals;
  const path = routes.comments(req.params.threadSlug);
  let collection = coll.base(path)(credentials);

  if (comments.length) {
    collection = R.compose(
      addCommentItems(res.locals)(comments)
    )(collection);
  }

  if (!policies.comments.post(req, res)) {
    return collection;
  }

  return R.compose(
    coll.withTemplate(['id', 'text'])({})
  )(collection);
};

/** The controller for the comments resource. */
module.exports = {
  get: (req, res) => {
    const { model } = res.locals;
    const { getCommentsByThreadSlug } = model.comments;
    const comments = getCommentsByThreadSlug(req.params.threadSlug);
    return createCommentsCollection({ req, res, comments });
  },
  post: (req, res) => {
    const { credentials, model, routes } = res.locals;
    const { createComment } = model.comments;
    const input = parseRequestBody(req);
    const comment = createComment({
      ...input,
      threadSlug: req.params.threadSlug,
    }, credentials);
    const location = href(routes.comment(comment.id));

    res.status(201)
      .set('Location', location)
      .set('Content-Location', location);

    return createCommentCollection(res.locals)(comment);
  },
};
