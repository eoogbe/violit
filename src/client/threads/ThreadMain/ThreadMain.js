/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useCallback, useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { ArticleFooter, ArticleHeader } from 'client/article';
import { CommentForm } from 'client/comments';
import { Markdown, PathSavingLink, useTitle } from 'client/core';
import * as api from '../api';

const Headline = styled.h2`
  font-size: 1.75rem;
  margin: 0;
`;

const NoArticleBodyParagraph = styled.p`
  font-size: 1rem;
  margin: ${({ theme }) => theme.spacingMd} 0;
  color: ${({ theme }) => theme.gray700};
`;

NoArticleBodyParagraph.displayName = 'NoArticleBodyParagraph';

const NoCommentFormParagraph = styled.p`
  font-size: 1rem;
  margin: ${({ theme }) => theme.spacingMd} 0;
  color: ${({ theme }) => theme.gray700};

  a {
    color: ${({ theme }) => theme.gray700};
    color: ${({ theme }) => theme.black}
  }
`;

/** The container for the main content of the thread screen. */
const ThreadMain = (props) => {
  const {
    id,
    slug,
    headline,
    articleBody,
    author,
    dateCreated,
    loggedIn,
    createComment,
    onDeleteError,
  } = props;
  const [dateDeleted, setDateDeleted] = useState(props.dateDeleted);
  const [
    canCreateComment,
    setCanCreateComment,
  ] = useState(props.canCreateComment);
  const [canDelete, setCanDelete] = useState(props.canDelete);

  const deleteThread = useCallback(async () => {
    setCanDelete(false);
    setCanCreateComment(false);
    setDateDeleted(new Date().toISOString());

    try {
      await api.sendDeleteThread(slug);
    } catch (err) {
      if (onDeleteError) {
        onDeleteError(err);
      }
    }
  }, [slug, onDeleteError]);

  useTitle(headline);

  return (
    <>
      <Headline itemProp="headline">{headline}</Headline>
      <ArticleHeader author={author} dateCreated={dateCreated} />
      <div itemProp="articleBody">
        {!dateDeleted && articleBody && <Markdown source={articleBody} />}
        {dateDeleted &&
          <NoArticleBodyParagraph>[deleted]</NoArticleBodyParagraph>}
      </div>
      {loggedIn && canCreateComment &&
        <CommentForm createComment={createComment} />}
      {!dateDeleted && !canCreateComment &&
        <NoCommentFormParagraph>
          What are your thoughts?&nbsp;
          <PathSavingLink to="/login">Log in</PathSavingLink>
        </NoCommentFormParagraph>}
      {loggedIn && canDelete &&
        <ArticleFooter
          contentId={`thread-${id}`}
          deleteContent={deleteThread}
        />}
    </>
  );
};

ThreadMain.propTypes = {
  /** The unique id of the thread. */
  id: PropTypes.string.isRequired,
  /** The url-optimized slug of the thread. */
  slug: PropTypes.string.isRequired,
  /** The username of the person who started the thread. */
  author: PropTypes.string.isRequired,
  /** The date the thread was posted. */
  dateCreated: PropTypes.string.isRequired,
  /** The date the thread was deleted. */
  dateDeleted: PropTypes.string,
  /** The title of the thread. */
  headline: PropTypes.string.isRequired,
  /** The body of the thread. */
  articleBody: PropTypes.string,
  /** `true` if the current user is logged in. */
  loggedIn: PropTypes.bool,
  /** `true` if the current user can create a comment. */
  canCreateComment: PropTypes.bool,
  /** Creates a comment. */
  createComment: PropTypes.func.isRequired,
  /** `true` if the current user can delete the thread. */
  canDelete: PropTypes.bool,
  /** Called when comment deletion fails. */
  onDeleteError: PropTypes.func,
};

export default ThreadMain;
