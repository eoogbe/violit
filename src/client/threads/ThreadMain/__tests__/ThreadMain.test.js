/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { MemoryRouter } from 'react-router-dom';
import TestRenderer from 'react-test-renderer';
import * as THREADS from '__fixtures__/threadsFixture';
import { ArticleFooter } from 'client/article';
import { CommentForm } from 'client/comments';
import { Input } from 'client/forms';
import mockFetch from 'testUtils/mockFetch';
import { updateWrapper, wait } from 'testUtils/wait';
import ThreadMain from '../ThreadMain';

const NOW = 1572566400000;

const simulateCommentFormSubmit = (wrapper) => {
  const onSubmit = wrapper.find(CommentForm).children().prop('onSubmit');
  return act(async () => {
    await onSubmit({
      preventDefault: () => {},
    });
    await updateWrapper(wrapper);
  });
};

it('render without crashing', () => {
  Date.now = jest.fn().mockReturnValue(NOW);
  const renderer = TestRenderer.create(
    <MemoryRouter>
      <ThreadMain
        id="thread1"
        slug={THREADS.thread1.slug}
        author={THREADS.thread1.author}
        dateCreated={THREADS.thread1.dateCreated}
        headline={THREADS.thread1.headline}
        articleBody={THREADS.thread1.articleBody}
        createComment={() => {}}
        loggedIn
        canCreateComment
        canDelete
      />
    </MemoryRouter>
  );
  expect(renderer.toJSON()).toMatchSnapshot();
});

it('renders without article body', () => {
  Date.now = jest.fn().mockReturnValue(NOW);
  const renderer = TestRenderer.create(
    <MemoryRouter>
      <ThreadMain
        id="thread1"
        slug={THREADS.thread1.slug}
        author={THREADS.thread1.author}
        dateCreated={THREADS.thread1.dateCreated}
        headline={THREADS.thread1.headline}
        createComment={() => {}}
        loggedIn
        canCreateComment
        canDelete
      />
    </MemoryRouter>
  );
  expect(renderer.toJSON()).toMatchSnapshot();
});

it('renders when deleted', () => {
  Date.now = jest.fn().mockReturnValue(NOW);
  const renderer = TestRenderer.create(
    <MemoryRouter>
      <ThreadMain
        id="thread1"
        slug={THREADS.thread1.slug}
        author={THREADS.thread1.author}
        dateCreated={THREADS.thread1.dateCreated}
        dateDeleted={THREADS.thread1.dateDeleted}
        headline={THREADS.thread1.headline}
        articleBody={THREADS.thread1.articleBody}
        createComment={() => {}}
      />
    </MemoryRouter>
  );
  expect(renderer.toJSON()).toMatchSnapshot();
});

it('renders when cannot create a comment', () => {
  Date.now = jest.fn().mockReturnValue(NOW);
  const renderer = TestRenderer.create(
    <MemoryRouter>
      <ThreadMain
        id="thread1"
        slug={THREADS.thread1.slug}
        author={THREADS.thread1.author}
        dateCreated={THREADS.thread1.dateCreated}
        headline={THREADS.thread1.headline}
        articleBody={THREADS.thread1.articleBody}
        createComment={() => {}}
      />
    </MemoryRouter>
  );
  expect(renderer.toJSON()).toMatchSnapshot();
});

it('sets the document title', () => {
  TestRenderer.act(() => {
    TestRenderer.create(
      <MemoryRouter>
        <ThreadMain
          id="thread1"
          slug={THREADS.thread1.slug}
          author={THREADS.thread1.author}
          dateCreated={THREADS.thread1.dateCreated}
          headline={THREADS.thread1.headline}
          articleBody={THREADS.thread1.articleBody}
          createComment={() => {}}
          loggedIn
          canCreateComment
          canDelete
        />
      </MemoryRouter>
    );
  });
  expect(document.title).toBe(THREADS.thread1.title);
});

it('when submitting comment form creates a comment', async () => {
  const createComment = jest.fn();

  const wrapper = mount(
    <MemoryRouter>
      <ThreadMain
        id="thread1"
        slug={THREADS.thread1.slug}
        author={THREADS.thread1.author}
        dateCreated={THREADS.thread1.dateCreated}
        headline={THREADS.thread1.headline}
        articleBody={THREADS.thread1.articleBody}
        createComment={createComment}
        loggedIn
        canCreateComment
        canDelete
      />
    </MemoryRouter>
  );

  wrapper.find(Input).simulate('change', {
    target: { name: 'text', value: 'Lorem ipsum' },
  });

  await simulateCommentFormSubmit(wrapper);

  expect(createComment).toHaveBeenCalled();
});

describe('when delete menu item clicked', () => {
  it('deletes the thread', async () => {
    mockFetch([
      THREADS.deleteThreadFetch,
    ]);

    const wrapper = mount(
      <MemoryRouter>
        <ThreadMain
          id="thread1"
          slug={THREADS.thread1.slug}
          author={THREADS.thread1.author}
          dateCreated={THREADS.thread1.dateCreated}
          headline={THREADS.thread1.headline}
          articleBody={THREADS.thread1.articleBody}
          createComment={() => {}}
          loggedIn
          canCreateComment
          canDelete
        />
      </MemoryRouter>
    );

    expect(wrapper.find('NoArticleBodyParagraph')).toHaveLength(0);
    expect(wrapper.find(CommentForm)).toHaveLength(1);
    expect(wrapper.find(ArticleFooter)).toHaveLength(1);

    wrapper.find('OverflowToggle').simulate('click');
    wrapper.find('MenuItemButton').first().simulate('click');

    expect(wrapper.find('NoArticleBodyParagraph')).toHaveLength(1);
    expect(wrapper.find(CommentForm)).toHaveLength(0);
    expect(wrapper.find(ArticleFooter)).toHaveLength(0);

    await wait();
  });
  it('when delete fails calls delete error callback', async () => {
    mockFetch([
      {
        url: THREADS.thread1.deleteUrl,
        method: 'POST',
        status: 500,
      },
    ]);

    const onDeleteError = jest.fn();
    const wrapper = mount(
      <MemoryRouter>
        <ThreadMain
          id="thread1"
          slug={THREADS.thread1.slug}
          author={THREADS.thread1.author}
          dateCreated={THREADS.thread1.dateCreated}
          headline={THREADS.thread1.headline}
          articleBody={THREADS.thread1.articleBody}
          createComment={() => {}}
          onDeleteError={onDeleteError}
          loggedIn
          canCreateComment
          canDelete
        />
      </MemoryRouter>
    );

    wrapper.find('OverflowToggle').simulate('click');
    wrapper.find('MenuItemButton').first().simulate('click');

    await wait();

    expect(onDeleteError).toHaveBeenCalled();
  });
  it('without delete error callback does nothing', async () => {
    mockFetch([
      {
        url: THREADS.thread1.deleteUrl,
        method: 'POST',
        status: 500,
      },
    ]);

    const wrapper = mount(
      <MemoryRouter>
        <ThreadMain
          id="thread1"
          slug={THREADS.thread1.slug}
          author={THREADS.thread1.author}
          dateCreated={THREADS.thread1.dateCreated}
          headline={THREADS.thread1.headline}
          articleBody={THREADS.thread1.articleBody}
          createComment={() => {}}
          loggedIn
          canCreateComment
          canDelete
        />
      </MemoryRouter>
    );

    wrapper.find('OverflowToggle').simulate('click');
    wrapper.find('MenuItemButton').first().simulate('click');

    await wait();
  });
});
