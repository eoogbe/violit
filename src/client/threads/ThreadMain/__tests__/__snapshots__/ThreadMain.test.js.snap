// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render without crashing 1`] = `
Array [
  <h2
    className="sc-hMqMXs hgQeqg"
    itemProp="headline"
  >
    Lorem ipsum
  </h2>,
  <header
    className="sc-gqjmRU dBUhVe"
  >
    <span
      className="sc-VigVT djDVvN"
      itemProp="author"
    >
      fred-astaire
    </span>
    <time
      dateTime="2019-09-29T01:23:45.679Z"
      itemProp="dateCreated"
      title="2019-09-29T01:23:45.679Z"
    >
      1 month ago
    </time>
  </header>,
  <div
    itemProp="articleBody"
  >
    <div
      className="sc-gZMcBi dDZOsr"
    >
      <p>
        Dolor sit amet, consectetur adipiscing elit.
      </p>
    </div>
  </div>,
  <form
    className="sc-cSHVUG jxHEP"
    onSubmit={[Function]}
  >
    <ul
      className="sc-kAzzGY ckrMOY"
    >
      <li>
        <textarea
          className="sc-jzJRlG hNkXwR"
          name="text"
          onChange={[Function]}
          placeholder="Write a comment"
          rows={5}
          value=""
        />
      </li>
    </ul>
    <button
      className="sc-dnqmqq bECuVm sc-chPdSV bBcFRQ"
      disabled={true}
      type="submit"
    >
      Comment
    </button>
  </form>,
  <footer
    className="sc-htpNat gPlOuz"
  >
    <nav
      className="sc-bxivhb eysMmV"
    >
      <button
        aria-controls="overflow-menu-thread-thread1"
        aria-expanded={false}
        aria-haspopup="true"
        className="sc-ifAKCX cLgJZS"
        onClick={[Function]}
        type="button"
      >
        <svg
          fill="currentColor"
          focusable={false}
          height="1.5em"
          size="1.5em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          title="More actions"
          viewBox="0 0 24 24"
          width="1.5em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
          />
        </svg>
      </button>
    </nav>
  </footer>,
]
`;

exports[`renders when cannot create a comment 1`] = `
Array [
  <h2
    className="sc-hMqMXs hgQeqg"
    itemProp="headline"
  >
    Lorem ipsum
  </h2>,
  <header
    className="sc-gqjmRU dBUhVe"
  >
    <span
      className="sc-VigVT djDVvN"
      itemProp="author"
    >
      fred-astaire
    </span>
    <time
      dateTime="2019-09-29T01:23:45.679Z"
      itemProp="dateCreated"
      title="2019-09-29T01:23:45.679Z"
    >
      1 month ago
    </time>
  </header>,
  <div
    itemProp="articleBody"
  >
    <div
      className="sc-gZMcBi dDZOsr"
    >
      <p>
        Dolor sit amet, consectetur adipiscing elit.
      </p>
    </div>
  </div>,
  <p
    className="sc-kkGfuU hDjocN"
  >
    What are your thoughts?Â 
    <a
      href="/login"
      onClick={[Function]}
    >
      Log in
    </a>
  </p>,
]
`;

exports[`renders when deleted 1`] = `
Array [
  <h2
    className="sc-hMqMXs hgQeqg"
    itemProp="headline"
  >
    Lorem ipsum
  </h2>,
  <header
    className="sc-gqjmRU dBUhVe"
  >
    <span
      className="sc-VigVT djDVvN"
      itemProp="author"
    >
      fred-astaire
    </span>
    <time
      dateTime="2019-09-29T01:23:45.679Z"
      itemProp="dateCreated"
      title="2019-09-29T01:23:45.679Z"
    >
      1 month ago
    </time>
  </header>,
  <div
    itemProp="articleBody"
  >
    <p
      className="sc-kEYyzF jLXdTC"
    >
      [deleted]
    </p>
  </div>,
]
`;

exports[`renders without article body 1`] = `
Array [
  <h2
    className="sc-hMqMXs hgQeqg"
    itemProp="headline"
  >
    Lorem ipsum
  </h2>,
  <header
    className="sc-gqjmRU dBUhVe"
  >
    <span
      className="sc-VigVT djDVvN"
      itemProp="author"
    >
      fred-astaire
    </span>
    <time
      dateTime="2019-09-29T01:23:45.679Z"
      itemProp="dateCreated"
      title="2019-09-29T01:23:45.679Z"
    >
      1 month ago
    </time>
  </header>,
  <div
    itemProp="articleBody"
  />,
  <form
    className="sc-cSHVUG jxHEP"
    onSubmit={[Function]}
  >
    <ul
      className="sc-kAzzGY ckrMOY"
    >
      <li>
        <textarea
          className="sc-jzJRlG hNkXwR"
          name="text"
          onChange={[Function]}
          placeholder="Write a comment"
          rows={5}
          value=""
        />
      </li>
    </ul>
    <button
      className="sc-dnqmqq bECuVm sc-chPdSV bBcFRQ"
      disabled={true}
      type="submit"
    >
      Comment
    </button>
  </form>,
  <footer
    className="sc-htpNat gPlOuz"
  >
    <nav
      className="sc-bxivhb eysMmV"
    >
      <button
        aria-controls="overflow-menu-thread-thread1"
        aria-expanded={false}
        aria-haspopup="true"
        className="sc-ifAKCX cLgJZS"
        onClick={[Function]}
        type="button"
      >
        <svg
          fill="currentColor"
          focusable={false}
          height="1.5em"
          size="1.5em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          title="More actions"
          viewBox="0 0 24 24"
          width="1.5em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
          />
        </svg>
      </button>
    </nav>
  </footer>,
]
`;
